svm/tests/integration_test.rs

```diff
@@ -2631,3 +2635,405 @@ fn svm_metrics_accumulation() {
         );
     }
 }
+
+// NOTE this could be moved to its own file in the future, but it requires a total refactor of the test runner
+mod balance_collector {
+    use {
+        super::*,
+        rand0_7::prelude::*,
+        solana_account::state_traits::StateMut,
+        solana_sdk::bpf_loader,
+        spl_generic_token::{token, token_2022},
+        test_case::test_case,
+    };
+
+    // this could be part of mock_bank but so far nothing but this uses it
+    static SPL_TOKEN_BYTES: &[u8] =
+        include_bytes!("../../program-test/src/programs/spl_token-3.5.0.so");
+
+    const STARTING_BALANCE: u64 = LAMPORTS_PER_SOL * 100;
+
+    // a helper for constructing a transfer instruction, agnostic over system/token
+    // it also pulls double duty as a record of what the *result* of a transfer should be
+    // so we can instantiate a Transfer, gen the instruction, change it to fail, change the record to amount 0
+    // and then the final test confirms the pre/post balances are unchanged with no special casing
+    #[derive(Debug, Default)]
+    struct Transfer {
+        from: Pubkey,
+        to: Pubkey,
+        amount: u64,
+    }
+
+    impl Transfer {
+        // given a set of users, picks two randomly and does a random transfer between them
+        fn new_rand(users: &[Pubkey]) -> Self {
+            let mut rng = rand0_7::thread_rng();
+            let mut users = users.to_vec();
+            let (i, from) = users.iter().enumerate().choose(&mut rng).unwrap();
+            let from = *from;
+            users.remove(i);
+            let to = *users.iter().choose(&mut rng).unwrap();
+            let amount = rng.gen_range(1, STARTING_BALANCE / 100);
+
+            Self { from, to, amount }
+        }
+
+        fn to_system_instruction(&self) -> Instruction {
+            system_instruction::transfer(&self.from, &self.to, self.amount)
+        }
+
+        fn to_token_instruction(&self, fee_payer: Pubkey) -> Instruction {
+            // true tokenkeg connoisseurs will note we shouldnt have to sign the sender
+            // we use a common account owner, the fee-payer, to conveniently reuse account state
+            // so why do we sign? to force the sender and receiver to be in a consistent order in account keys
+            // which means we can grab them by index in our final test instead of searching by key
+            Instruction {
+                program_id: token::id(),
+                accounts: vec![
+                    AccountMeta::new(self.from, true),
+                    AccountMeta::new(self.to, false),
+                    AccountMeta::new(fee_payer, true),
+                ],
+                data: bincode::serialize(&(3u8, self.amount)).unwrap(),
```

Comment:
is there a constant 3 we can pull in from anywhere reasonable here?

Reply:
ill define it explicitly in the file, the only place to get it would be spl-token

Reply:
[c9a557c](https://github.com/anza-xyz/agave/pull/5588/commits/c9a557c9d33494c80ada35ea6d677814ce89af50)